Copy found blocked 3rd party mods to the new instance (ftb and curseforge)

Ryex opened the pull request:
<!--
Hey there! Thanks for your contribution.

Please make sure that your commits are signed off first.
If you don't know how that works, check out our contribution guidelines: https://github.com/PrismLauncher/PrismLauncher/blob/develop/CONTRIBUTING.md#signing-your-work
If you already created your commits, you can run `git rebase --signoff develop` to retroactively sign-off all your commits and `git push --force` to override what you have pushed already.

Note that signing and signing-off are two different things!
-->
This PR enhances the user experience when dealing with mods blocked from 3rd party launchers.

Fixes #222

When the BlockedMods dialog is created it check the configured global mods folder and the users' downloads folder for the missing .jar names and if found verifies them by hash. 
A filesystem watch is then established on both folders in case the user needs to download the missing mods.

The UI updates appropriately to indicate the blocked mods have been found.
Once the dialog has been accepted the mods are copied to the instance's mods folder.

Of note:

* UI is not necessarily final. 
* There are a few flow questions to answer Namely, 
  - should the dialog's OK button be disabled until all blocked mods are detected?
  - If blocked mod fails to copy for any reason should the whole instance creation be aborted?

Code review is needed.

Ryex committed:
send blocked mod info to dialog & prototype UI

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
watch filesystem, compute and match hashes

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
copy found mods to instance (FTB and Flame)

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln commented on launcher/modplatform/modpacksch/FTBPackInstallTask.cpp:
```diff
@@ -191,11 +189,19 @@ void PackInstallTask::onResolveModsSucceeded()
         // First check for blocked mods
         if (!results_file.resolved || results_file.url.isEmpty()) {
-            QString type(local_file.type);
+            // QString type(local_file.type);
+
+            // type[0] = type[0].toUpper();
+
```
```suggestion
```

Ryex replied:
cleaned up after myself. also added some doc comments

flowln commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,185 @@
+#include <qfileinfo.h>
+#include <qnamespace.h>
```
```suggestion
```

flowln commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,185 @@
+        << ", localPath: " << m.localPath <<"}";
+
+    return debug;
+}
```
add newline :o

flowln reviewed (commented):
just some preliminary stuff :)

Ryex committed:
minor clean up and add some docs

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -162,6 +162,35 @@ bool ensureFolderPathExists(QString foldernamepath)
+/// @param src srouce file path
+/// @param dst destination file path
+/// @return boolean: was there an error during the filecopy?
+bool copyFile(QString &src, QString &dst) {
```
I've fixed the issue of `FS::copy` not copying individual files in #333, so this can go once that is merged :p

Ryex replied:
Good to know, I've added a `FIXME\TODO` comment. Once #333 is merged I'll make the change.

flowln commented on launcher/modplatform/flame/FlameInstanceCreationTask.cpp:
```diff
@@ -404,6 +411,38 @@ void FlameCreationTask::idResolverSucceeded(QEventLoop& loop)
 
+/// @brief copy the matched blocked mods to the instance staging area
+/// @param blocked_mods list of the blocked mods and their matched paths
+void FlameCreationTask::copyBlockedMods(QList<BlockedMod> blocked_mods) {
```
```suggestion
void FlameCreationTask::copyBlockedMods(QList<BlockedMod> const& blocked_mods) {
```

note that this may create problems with the iterators you're using inside this function. however, you're not really mutating the `QList`, and this way we'll prevent a deep copy from happening if you access it in a non-const way (also, it may be better to use a range-for loop instead of a Java style one down there)

Ryex replied:
good catch. I actually add it marked as `const` at one point but had to remove it during debugging for a reason I forgot, it's no longer necessary not to now though.

flowln commented on launcher/modplatform/flame/FlameInstanceCreationTask.cpp:
```diff
@@ -449,7 +488,9 @@ void FlameCreationTask::setupDownloadJob(QEventLoop& loop)
         setError(reason);
     });
-    connect(m_files_job.get(), &NetJob::progress, [&](qint64 current, qint64 total) { setProgress(current, total); });
+    connect(m_files_job.get(), &NetJob::progress, [&](qint64 current, qint64 total) { 
+        setProgress(current, total); 
+    });
```
```suggestion
    connect(m_files_job.get(), &NetJob::progress, this, &FlameCreationTask::setProgress);
```

flowln commented on launcher/modplatform/helpers/HashUtils.cpp:
```diff
@@ -35,6 +35,18 @@ Hasher::Ptr createFlameHasher(QString file_path)
+
+Hasher::Ptr createBlockedModHasher(QString file_path, ModPlatform::Provider provider, QString type)
+{
+    auto hasher = new BlockedModHasher(file_path, provider);
```
```suggestion
    auto hasher = createBlockedModHasher(file_path, provider);
```

this way we avoid having to change two places at once if we need to change something in there :)

Ryex replied:
the `useHashType` method used to pass along the specific hash type isn't part of the `Hasher::ptr` interface so the type narrowing that results from using the builder prevents it's use.

flowln replied:
eh ok then, it's kinda weird that the previous class hierarchy uses subclassing for different providers, while here we use type definition to change this behavior. welp

flowln commented on launcher/modplatform/modpacksch/FTBPackInstallTask.cpp:
```diff
@@ -320,6 +326,9 @@ void PackInstallTask::downloadPack()
 void PackInstallTask::onModDownloadSucceeded()
 {
     m_net_job.reset();
+    if (m_blocked_mods.length() > 0) {
```
```suggestion
    if (!m_blocked_mods.isEmpty()) {
```

Ryex replied:
... I was looking for that method but for some reason my editor told me it didn't exist the first time....

flowln commented on launcher/modplatform/modpacksch/FTBPackInstallTask.cpp:
```diff
@@ -343,4 +352,35 @@ void PackInstallTask::onModDownloadFailed(QString reason)
+    int i = 0;
+    int total = m_blocked_mods.length();
+    setProgress(i, total);
+    for (auto mod = m_blocked_mods.begin(); mod != m_blocked_mods.end(); mod++, i++) {
```
a range-for loop would be nicer here i think

Ryex replied:
done, I was being silly when I used it just to attach `i` to the loop.

flowln commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,193 @@
     this->setWindowTitle(title);
     ui->label->setText(text);
-    ui->textBrowser->setText(body);
+    ui->labelModsFound->setText("Please download the missing mods.");
```
```suggestion
    ui->labelModsFound->setText(tr("Please download the missing mods."));
```

for translations :>

flowln commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,193 @@
+}
+
+/// @brief update UI with current status of the blocked mod detection
+void BlockedModsDialog::update() {
```
add `tr()` to the user-facing english text in this function pls

Ryex replied:
Is there a place I need to add the new strings for them to get added to Weblate when this is merged?

flowln replied:
nop, just adding `tr()` is enough

Ryex replied:
Ok then, we should be all good on that then.

flowln commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,193 @@
+
+}
+
+/// @brief check if the conputed hash for the provided path matches a blocked
```
```suggestion
/// @brief check if the computed hash for the provided path matches a blocked
```

flowln commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,193 @@
     }
 }
+
+/// @brief Check if the name of the file at path matches the naem of a blocked mod we are searching for
```
```suggestion
/// @brief Check if the name of the file at path matches the name of a blocked mod we are searching for
```

flowln commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,193 @@
+    return false;
+}
+
+/// @brief have we found all the mods we're lookign for?
```
```suggestion
/// @brief have we found all the mods we're looking for?
```

flowln commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,193 @@
+}
+
+/// @brief have we found all the mods we're lookign for?
+/// @return boolean
```
well i can see the function signature, no need for this comment if it's not gonna tell me anything xD

Ryex replied:
fair enough

flowln commented on launcher/ui/dialogs/BlockedModsDialog.h:
```diff
@@ -11,12 +27,27 @@ class BlockedModsDialog : public QDialog {
+    bool allModsMatched();
 };
+
+QDebug operator<<(QDebug debug, const BlockedMod &m);
```
can't mark the thing, but add newline!!!

Ryex committed:
fix mispellings and wrap strings for translation

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
code quality cleanup

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
ensure FS::copyFile is marked for removal

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
chore(remove FS::copyFile):
Now that #333 is merged and FS::copy works on non directory copyFile can be removed.

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
Rebased and removed `FS::copyFile` now that #333 is merged with it's fix for `FS::copy`

FederAndInk commented on launcher/modplatform/flame/FlameInstanceCreationTask.cpp:
```diff
@@ -404,6 +411,39 @@ void FlameCreationTask::idResolverSucceeded(QEventLoop& loop)
+    int i = 0;
+    int total = blocked_mods.length();
+    setProgress(i, total);
+    for (auto &mod : blocked_mods) {
```
you might want to run clang-format on your modifications, there are some inconsistencies in the formatting

FederAndInk replied:
you can also add the const there ^^
```suggestion
    for (auto const& mod : blocked_mods) {
```

FederAndInk commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,191 @@
+            mod.localPath = path;
+            match = true;
+
+            qDebug() << "Hash match found: " << mod.name << " " << hash << " | From path:" << path;
```
`qDebug() <<` adds spaces automatically around `<<` :slightly_smiling_face: 

```suggestion
            qDebug() << "Hash match found:" << mod.name << hash << "| From path:" << path;
```

FederAndInk commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,191 @@
+}
+
+bool BlockedModsDialog::allModsMatched() {
+    for (auto &mod : mods) {
+        if (!mod.matched)
+            return false;
+    }
+    return true;
```
You can use `std::all_of` if you want ^^
https://en.cppreference.com/w/cpp/algorithm/all_any_none_of

Ryex replied:
ty

FederAndInk replied:
no problem :)

FederAndInk commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,191 @@
+        connect(hash_task.get(), &Task::succeeded, [this, hash_task, file] { 
+            checkMatchHash(hash_task->getResult(), file);
+        });
+        connect(hash_task.get(), &Task::failed, [this, hash_task, file] { 
+            qDebug() << "Failed to hash path: " << file;
+        });
```
You can remove `this, hash_task` from the capture list

Ryex committed:
clang_format and code cleanup

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln reviewed (approved):
looks good to me, thanks!

Scrumplex commented on launcher/modplatform/modpacksch/FTBPackInstallTask.cpp:
```diff
@@ -210,13 +213,16 @@ void PackInstallTask::onResolveModsSucceeded()
 
-        if (message_dialog->exec() == QDialog::Accepted)
+        if (message_dialog->exec() == QDialog::Accepted) {
+            qDebug() << "Post dialog mods list: " << m_blocked_mods;
```
Maybe this should be added to CF instance creation task. Or remove it here

Ryex replied:
This was mostly used during the initial debugging but it seems useful enough to keep in the logs at the time. I forgot that I hadn't put it on the Flame creation task too. Would you prefer it removed or duplicated for consistency?

Scrumplex replied:
I think it would be nice for debugging, so sure, go ahead and add it to Flame as well :D

Scrumplex commented on launcher/modplatform/modpacksch/FTBPackInstallTask.cpp:
```diff
@@ -343,4 +352,35 @@ void PackInstallTask::onModDownloadFailed(QString reason)
     emitFailed(reason);
 }
 
+/// @brief copy the matched blocked mods to the instance staging area
+void PackInstallTask::copyBlockedMods()
+{
+    setStatus(tr("Copying Blocked Mods..."));
+    setAbortable(false);
+    int i = 0;
+    int total = m_blocked_mods.length();
+    setProgress(i, total);
+    for (auto const& mod : m_blocked_mods) {
+        if (!mod.matched) {
+            qDebug() << mod.name << "was not matched to a local file, skipping copy";
+            continue;
+        }
+
+        auto dest_path = FS::PathCombine(m_stagingPath, ".minecraft", "mods", mod.name);
+
+        setStatus(tr("Copying Blocked Mods (%1 out of %2 are done)").arg(QString::number(i), QString::number(total)));
+
+        qDebug() << "Will try to copy" << mod.localPath << "to" << dest_path;
+
+        if (!FS::copy(mod.localPath, dest_path)()) {
+            qDebug() << "Copy of" << mod.localPath << "to" << dest_path << "Failed";
+        }
+
+        i++;
+        setProgress(i, total);
+    }
+
+    setAbortable(true);
+}
```
Duplicate code, but I guess there is no better way currently.

Ryex replied:
yeah, because there is a significant difference between how the FTB and normal Flame install tasks go about things I couldn't think of a clean way to de-duplicate this code.

Scrumplex commented on launcher/ui/dialogs/BlockedModsDialog.cpp:
```diff
@@ -1,28 +1,186 @@
+
+/// @brief Signal fired when a watched direcotry has changed
+/// @param path the path to the changed directory
+void BlockedModsDialog::directoryChanged(QString path)
+{
+    qDebug() << "Directory changed: " << path;
+    scanPath(path);
+}
```
This could cause severe slowdown if the whole directory is being re-scanned every time a new file is added :thinking:

Ryex replied:
There whole directory is not re-scanned, only files that match the right filenames have their hash's checked.
the scan effectively becomes loop over the current filenames. even if there was an attempt to detect which files are "new" since the last check this loop would need to be done. additionally some browsers do not use a part file and download the file directly to the target name and so we have to recheck even files we checked before if their contents change.

Scrumplex reviewed (approved):
Looks good overall. Just a few smaller things.

Ryex committed:
add some extra debug logs for CF blocked mods

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
We all good now :+1:

DioEgizio commented:
for people with a custom download location, would be cool if you could just drag and drop or browse and insert the mods

Ryex commented:
> for people with a custom download location, would be cool if you could just drag and drop or browse and insert the mods

These are good suggestions and relatively easy to add, would you prefer them in this PR or a separate one?

flowln commented:
> These are good suggestions and relatively easy to add, would you prefer them in this PR or a separate one?

Probably another PR would be better, this one already has a good amount of changes, and will ~soon~ be merged :>
